### PR TITLE
Cleanup the number of helper methods that are available for all specs

### PR DESCRIPTION
**Pull request by GustavoCaso**
**What does this PR do?**

Clean up our `spec_helper` file and reduce the number of helpers included for all specs. Also, delete duplicated shared_exmaples, and unused custom matchers.

If a few specs need a helper, we should first require the spec helper directly, rather than adding as part of the global available helpers. Once we identify that a helper is widely used we can promote it to be part of all the specs. This will keep the `spec_helper.rb` lean, and it would be easier to understand what is that all the specs need in order to work properly. 

**Motivation:**

**Additional Notes:**

**How to test the change?**

**For Datadog employees:**
- [ ] If this PR touches code that signs or publishes builds or packages, or handles
credentials of any kind, I've requested a review from `@DataDog/security-design-and-guidance`.
- [ ] This PR doesn't touch any of that.

Unsure? Have a question? Request a review!

### CODE REVIEW

File: spec/datadog/tracing/buffer_spec.rb [hunk 1]
```diff
@@ -22,169 +23,6 @@
   end
 end
 
-RSpec.shared_examples 'thread-safe buffer' do
```
Comment (ivoanjo) [inline spec/datadog/tracing/buffer_spec.rb:25]: Weird that this was already duplicated in `buffer/shared_examples`...

**Review by ivoanjo — APPROVED** [review]
:+1: Seems reasonable, thanks for the cleanup!

File: spec/datadog/core/buffer/shared_examples.rb [hunk 2]
```diff
@@ -61,7 +61,14 @@
...
-          allow(Datadog).to receive(:logger).and_return(double)
+          logger = double(Datadog::Core::Logger)
+          allow(logger).to receive(:debug?).and_return true
+          allow(logger).to receive(:debug)
+          allow(logger).to receive(:info)
+          allow(logger).to receive(:warn)
+          allow(logger).to receive(:error)
+
+          allow(Datadog).to receive(:logger).and_return(logger)
         end
```
[the remark this answers is not in the document]
Comment (GustavoCaso) [reply spec/datadog/core/buffer/shared_examples.rb:72]: The change was done to replicate the shared example in `spec/datadog/tracing/buffer_spec.rb`. My two cents teel me that the logger in one of the shared examples was mocking fewer methods and couldn't be used for the `spec/datadog/tracing/buffer_spec.rb` file, so it got copied and modified to make it pass for that case.

File: spec/support/container_helpers.rb [hunk 3]
```diff
@@ -1,7 +1,7 @@
 require 'stringio'
```
[the remark this answers is not in the document]
Comment (GustavoCaso) [reply spec/support/container_helpers.rb:1]: That is already the case @ivoanjo 
image

https://github.com/DataDog/dd-trace-rb/pull/3173/files#diff-89eebfcbc0f14b6d989517837ca1e94fce4e2ce9a03233641cd936f2b8d2ed94L65

I guess the last remaining thing is move it to the `spec/datadog/core/environment/` folder

[the remark this answers is not in the document]
Comment (ivoanjo) [reply spec/support/container_helpers.rb:1]: > That is already the case @ivoanjo

Well, not entirely. As long as you require those specs, the helper will get required and made available for every other spec. But now that I look at them a bit closer, I think we'd need to to further changes if we wanted them to not "leak" to other specs.

So maybe just the move? ;)

[the remark this answers is not in the document]
Comment (GustavoCaso) [reply spec/support/container_helpers.rb:1]: I see want you mean 😄

[the remark this answers is not in the document]
Comment (GustavoCaso) [reply spec/support/container_helpers.rb:1]: I'll what we can do to cleanup the global helpers

**Pull request merged**